clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java

```diff
@@ -0,0 +1,476 @@
+ * This would result in a topic partition being sent back to the Lookup stage.
+ *
+ * Managing this complexity by chaining together `Call` implementations is challenging
+ * and messy, so instead we use this class to do the bookkeeping. It handles both the
+ * batching aspect as well as the transitions between the Lookup and Fulfillment stages.
+ *
+ * Note that the interpretation of the `retries` configuration becomes ambiguous
+ * for this kind of pipeline. We could treat it as an overall limit on the number
+ * of requests that can be sent, but that is not very useful because each pipeline
+ * has a minimum number of requests that need to be sent in order to satisfy the request.
+ * Instead, we treat this number of retries independently at each stage so that each
+ * stage has at least one opportunity to complete. So if a user sets `retries=1`, then
+ * the full pipeline can still complete as long as there are no request failures.
+ *
+ * @param <K> The key type, which is also the granularity of the request routing (e.g.
+ *            this could be `TopicPartition` in the case of requests intended for a partition
+ *            leader or the `GroupId` in the case of consumer group requests intended for
+ *            the group coordinator)
+ * @param <V> The fulfillment type for each key (e.g. this could be consumer group state
+ *            when the key type is a consumer `GroupId`)
+ */
+public class AdminApiDriver<K, V> {
+    private final Logger log;
+    private final long retryBackoffMs;
+    private final long deadlineMs;
+    private final AdminApiHandler<K, V> handler;
+    private final Optional<StaticKeyMapping<K>> staticMapping;
+    private final Optional<DynamicKeyMapping<K>> dynamicMapping;
+    private final Map<K, KafkaFutureImpl<V>> futures;
+
+    private final BiMultimap<RequestScope, K> lookupMap = new BiMultimap<>();
+    private final BiMultimap<BrokerScope, K> fulfillmentMap = new BiMultimap<>();
+    private final Map<RequestScope, RequestState> requestStates = new HashMap<>();
+
+
+    public AdminApiDriver(
+        AdminApiHandler<K, V> handler,
+        long deadlineMs,
+        long retryBackoffMs,
+        LogContext logContext
+    ) {
+        this.handler = handler;
+        this.deadlineMs = deadlineMs;
+        this.retryBackoffMs = retryBackoffMs;
+        this.log = logContext.logger(AdminApiDriver.class);
+        this.futures = new HashMap<>();
+
+        KeyMappings<K> result = handler.initializeKeys();
+        this.dynamicMapping = result.dynamicMapping;
+        this.staticMapping = result.staticMapping;
+
+        initializeKeys();
+    }
+
+    private void initializeKeys() {
+        staticMapping.ifPresent(mapping -> {
+            for (Map.Entry<K, Integer> staticMapping : mapping.keys.entrySet()) {
+                K key = staticMapping.getKey();
+                Integer brokerId = staticMapping.getValue();
```

Comment:
nit: Could we use `mapping.keys.forEach((key, brokerId) -> ..)`? It is a bit more concise.